management/src/main/java/io/micronaut/management/endpoint/env/EnvironmentFilterSpecification.java

```diff
@@ -0,0 +1,121 @@
+ *
+ * @author Tim Yates
+ * @since 3.3.0
+ */
+public class EnvironmentFilterSpecification {
+
+    private static final String[] PROPERTY_NAMES_TO_MASK = new String[]{
+            "password", "credential", "certificate", "key", "secret", "token"
+    };
+
+    @Nullable
+    private final Principal principal;
+    private boolean allMasked;
+    private final List<Pattern> maskedPatterns = new ArrayList<>();
+    private final List<Pattern> allowedPatterns = new ArrayList<>();
+
+    EnvironmentFilterSpecification(@Nullable Principal principal) {
+        this.principal = principal;
+        this.allMasked = true;
+    }
+
+    /**
+     * @return The current {@link Principal} that is making the request (if any)
+     */
+    @Nullable
+    public Principal getPrincipal() {
+        return principal;
+    }
+
+    /**
+     * Turn on global masking. Items can be unmasked by adding their {@link Pattern} to {@link EnvironmentFilterSpecification#maskPatterns(Pattern...)}.
+     *
+     * @return itself for chaining calls.
+     */
+    public EnvironmentFilterSpecification maskAll() {
+        allMasked = true;
+        return this;
+    }
+
+    /**
+     * Turn off global masking. Items can be masked by adding their {@link Pattern} to {@link EnvironmentFilterSpecification#maskPatterns(Pattern...)} (Pattern...)}.
+     *
+     * @return itself for chaining calls.
+     */
+    public EnvironmentFilterSpecification maskNone() {
+        allMasked = false;
+        return this;
+    }
+
+    /**
+     * Adds patterns to the list of known mask patterns.  If the {@link EnvironmentFilterSpecification#maskAll()} flag is set
+     * then this list will be used for exceptions (values in clear-text).  If the {@link EnvironmentFilterSpecification#maskNone()}
+     * flag is set then this list will be those values that are masked.
+     *
+     * @param propertySourceNameMask one or more Patterns to match against property names for masking.
+     * @return itself for chaining calls.
+     */
+    public EnvironmentFilterSpecification maskPatterns(Pattern... propertySourceNameMask) {
+        maskedPatterns.addAll(Arrays.asList(propertySourceNameMask));
+        return this;
+    }
+
+    /**
+     * Adds the masking rules from legacy pre 3.3.0.  Effectively the same as calling {@link EnvironmentFilterSpecification#maskNone()}
+     * and adding patterns via {@link EnvironmentFilterSpecification#maskPatterns(Pattern...)} to mask anything containing the words
+     * {@code password, credential, certificate, key, secret} or {@code token}.
+     *
+     * @return itself for chaining calls.
+     */
+    public EnvironmentFilterSpecification legacyMasking() {
+        allMasked = false;
+        maskedPatterns.addAll(
+                Arrays.stream(PROPERTY_NAMES_TO_MASK)
+                        .map(s -> Pattern.compile(".*" + s + ".*", Pattern.CASE_INSENSITIVE))
+                        .collect(Collectors.toList())
+        );
+        return this;
+    }
```

Comment:
Are we ok compiling these patterns each request?  We could compile them at startup, or could use something like an AtomicReference compile them on demand and cache them

Reply:
Makes sense to compile then on startup

Reply:
Done.  I used a Stream to compile them all.  Just a plain repetitive array might be quicker, but with only 6 entries, I'm not sure by how much...